### PR TITLE
SI-2390-create-mtree-log-root

### PR DESCRIPTION
**Pull request by Allyson-English**
- Add merkle tree generation to the rewards issuance process and log the root 
      - tree is generated after checks so it contains all devices that will earn for given week 
- in order to fulfill the requirements doc, this still needs the vc signature to be added (which means updating the grpc response from the devices api, I'll post a link to that pr here once I've done that)

Notes:
- There's several things we might want to store:
      - root and week number
      - proofs for each user 

Included below is the output of the test and the solidity contract and hardhat code I used to verify it

Contract: 
```solidity
// SPDX-License-Identifier: UNLICENSED
pragma solidity ^0.8.9;

import "node_modules/@openzeppelin/contracts/utils/cryptography/MerkleProof.sol";

contract Simple {
    constructor() {}

    function Verify(
        bytes32[] calldata _merkleProof,
        bytes32 storageMerkleRoot,
        bytes32 data
    ) public pure returns (bool) {
        return MerkleProof.verify(_merkleProof, storageMerkleRoot, data);
    }
}

```

Hardhat tests:
```javascript

  describe("Deployment", function () {
      it("rewards test", async function () {
      const { simple } = await loadFixture(deployOneYearLockFixture);
      expect(
        await simple.Verify(
          [
            "0x24afd3eb42056733b8fcdc968f4cd5053a21f73b45c243c9aec02acedd5021f6",
            "0x44a910b6135806f28d52453051bef92ad728dafc6e9620c52d1af2cdeb2bf998",
          ], //proof
          "0x06a0f4216bffb19713864b1f3020b3a7e505a4bcbe95e5df74a288189cb49e17", //root
          "0x02bf4c8c4f1649d724c6c06bdbce5978817a336f732403fd7ddd5116b562f47e" //data
        )
      ).to.equal(true);

      expect(
        await simple.Verify(
          [
            "0x02bf4c8c4f1649d724c6c06bdbce5978817a336f732403fd7ddd5116b562f47e",
            "0x44a910b6135806f28d52453051bef92ad728dafc6e9620c52d1af2cdeb2bf998",
          ], //proof
          "0x06a0f4216bffb19713864b1f3020b3a7e505a4bcbe95e5df74a288189cb49e17", //root
          "0x24afd3eb42056733b8fcdc968f4cd5053a21f73b45c243c9aec02acedd5021f6" //data
        )
      ).to.equal(true);

      expect(
        await simple.Verify(
          [
            "0x7fa502ce4199a9414e6df513252d1cfdb5cc9c99a2a8728ffbb52f54744dab83",
            "0xddc0e73f2c7000c4204febcafcd5ff181da1c02094b50243750ee2b3f83fd0e8",
          ], //proof
          "0x06a0f4216bffb19713864b1f3020b3a7e505a4bcbe95e5df74a288189cb49e17", //root
          "0x2c1a64fdb4f64cc638bc80466f2b7492a29d6359f5f490dd71eb6e050731c486" //data
        )
      ).to.equal(true);

      expect(
        await simple.Verify(
          [
            "0x2c1a64fdb4f64cc638bc80466f2b7492a29d6359f5f490dd71eb6e050731c486",
            "0xddc0e73f2c7000c4204febcafcd5ff181da1c02094b50243750ee2b3f83fd0e8",
          ], //proof
          "0x06a0f4216bffb19713864b1f3020b3a7e505a4bcbe95e5df74a288189cb49e17", //root
          "0x7fa502ce4199a9414e6df513252d1cfdb5cc9c99a2a8728ffbb52f54744dab83" //data
        )
      ).to.equal(true);
    });
  });

```

### CODE REVIEW

File: internal/services/attestor.go [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+		codeBytes, err := abiEncode(dataTypes, d)
+		if err != nil {
+			w.log.Err(err).Msg("failed to abi encode attestation data")
+		}
+		encodedLeaves = append(encodedLeaves, crypto.Keccak256(crypto.Keccak256(codeBytes)))
+	}
+
+	// sort values for OpenZeppelin compatibility
+	sort.Slice(encodedLeaves, func(i, j int) bool {
+		return string(encodedLeaves[i]) < string(encodedLeaves[j])
```
Comment (elffjs) [inline internal/services/attestor.go:78]: Don't we want to byte sort?

Comment (elffjs) [reply internal/services/attestor.go:78]: https://pkg.go.dev/bytes#Compare

File: internal/services/attestor.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+				return crypto.Keccak256(data), nil
+			},
+			Mode:               mt.ModeProofGenAndTreeBuild,
+			DisableLeafHashing: true, // also required for OpenZeppelin compatilibity
+		},
+		log: logger,
+	}
+}
+
+func (a *Attestor) GenerateMerkleTree(data map[string]map[string]interface{}, dataTypes []string) (*mt.MerkleTree, error) {
```
Comment (elffjs) [inline internal/services/attestor.go:65]: This is a very strange signature.

File: internal/services/attestor.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+func NewAttestor(logger *zerolog.Logger) *Attestor {
+	return &Attestor{
+		config: mt.Config{
+			SortSiblingPairs: true, // parameter for OpenZeppelin compatibility
+			HashFunc: func(data []byte) ([]byte, error) {
+				return crypto.Keccak256(data), nil
+			},
+			Mode:               mt.ModeProofGenAndTreeBuild,
+			DisableLeafHashing: true, // also required for OpenZeppelin compatilibity
```
Comment (elffjs) [inline internal/services/attestor.go:59]: Both here and on line 22, the point is that we're already serializing and hashing, yeah?

File: internal/services/rewards.go [hunk 4]
```diff
@@ -292,11 +296,40 @@ func (t *BaselineClient) assignPoints() error {
...
+						Values:              []interface{}{*ud.TokenId, ud.DeviceDefinitionId, ud.LatestVinCredential.Id},
+					}
+				}
+
+			}
+		}
+		attestationData[*ud.Vin] = map[string]interface{}{
+			CredentialExpiresAt: ud.LatestVinCredential.Expiration.AsTime(),
+			Values:              []interface{}{*ud.TokenId, ud.DeviceDefinitionId, ud.LatestVinCredential.Id},
+		}
```
Comment (elffjs) [inline internal/services/rewards.go:316]: Could these all be structs? I don't see the advantage of passing all these maps and arrays around.

File: internal/services/rewards.go [hunk 5]
```diff
@@ -292,11 +296,40 @@ func (t *BaselineClient) assignPoints() error {
...
+
 		if err := thisWeek.Insert(ctx, t.TransferService.db.DBS().Writer, boil.Infer()); err != nil {
```
Comment (elffjs) [inline internal/services/rewards.go:318]: I think the thing is we no longer know that every row in here is going to earn. I understand that we're not trying to implement _everything_ in this PR but I'm not sure where the responsibilities end and begin.

File: internal/services/attestor.go (hunk 3 above)
Comment (Allyson-English) [reply internal/services/attestor.go:59]: yes-- updated the doc string to clarify this. I'm wondering if I should try making a pr to the `"github.com/txaty/go-merkletree"` library

File: internal/services/rewards.go (hunk 4 above)
Comment (Allyson-English) [reply internal/services/rewards.go:316]: fair point, updated this to be a struct which also made the func signature of generateMerkleTree less odd

**Pull request closed without merging**